packages/component/src/Activity2/Timestamp.js

```diff
@@ -0,0 +1,83 @@
+import { connect } from 'react-redux';
+import React from 'react';
+
+import String, { getString } from '../Localization/String';
+import Context from '../Context';
+import TimeAgo from '../Utils/TimeAgo';
+
+// TODO: We could refactor this into a general component
+function sendFailed(language, replace) {
+  const text = getString('Send failed, {retry}', language);
+  const retry = getString('retry', language);
+  const match = /\{retry\}/.exec(text);
+
+  if (match) {
+    return (
+      <React.Fragment>
+        { text.substr(0, match.index) }
+        { replace(retry) }
+        { text.substr(match.index + match[0].length) }
+      </React.Fragment>
+    );
+  } else {
+    return text;
+  }
+}
+
+class Timestamp extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.handleRetryClick = this.handleRetryClick.bind(this);
+  }
+
+  handleRetryClick(evt) {
+    evt.preventDefault();
+
+    this.props.postActivity(this.props.activity);
+
+    // After clicking on "retry", the button will be gone and focus will be lost (back to document.body)
+    // We want to make sure the user stay inside Web Chat
```

Comment:
should this be "onRetryClick()" and the parent can decide to do the focus thing? The child shouldn't know about the send box.

Reply:
For customizability, we use a Context/Composer object to allow every DOM elements to have programmatic access to Web Chat APIs (e.g. `postActivity`, `startSpeechInput`, `submitSendBox`).

You shared a good insight. If the user is sending two succeeding bubbles within 5 minutes, we should show the retry button (and currently, we hide the timestamp, which double as a retry button).

Will fix.